lib/modules/platform/bitbucket-server/utils.ts

```diff
@@ -156,39 +158,93 @@ export function getInvalidReviewers(err: BitbucketError): string[] {
   return invalidReviewers;
 }
 
+function generateUrlFromEndpoint(
+  defaultEndpoint: string,
+  opts: HostRule,
+  repository: string
+): string {
+  const url = new URL(defaultEndpoint);
+  const generatedUrl = git.getUrl({
+    protocol: url.protocol as GitProtocol,
+    auth: `${opts.username}:${opts.password}`,
+    host: `${url.host}${url.pathname}${
+      // eslint-disable-next-line @typescript-eslint/no-unnecessary-type-assertion
+      url.pathname!.endsWith('/') ? '' : /* istanbul ignore next */ '/'
+    }scm`,
+    repository,
+  });
+  logger.debug({ url: generatedUrl }, `using generated endpoint URL`);
+  return generatedUrl;
+}
+
 export function getRepoGitUrl(
   repository: string,
   defaultEndpoint: string,
+  gitUrl: GitUrlOption | undefined,
   info: BbsRestRepo,
   opts: HostRule
-): string {
-  let cloneUrl = info.links.clone?.find(({ name }) => name === 'http');
-  if (!cloneUrl) {
-    // Http access might be disabled, try to find ssh url in this case
-    cloneUrl = info.links.clone?.find(({ name }) => name === 'ssh');
-  }
+): string | null {
+  switch (gitUrl) {
+    case 'default': {
+      const httpUrl = info.links.clone?.find(({ name }) => name === 'http');
+      if (httpUrl === undefined) {
+        return null;
+      }
+
+      logger.debug({ url: httpUrl.href }, `using http URL`);
+      // Inject auth into the API provided URL
+      const repoUrl = url.parse(httpUrl.href);
+      repoUrl.auth = `${opts.username}:${opts.password}`;
+      return url.format(repoUrl);
+    }
+    case 'ssh': {
+      const sshUrl = info.links.clone?.find(({ name }) => name === 'ssh');
+      if (sshUrl === undefined) {
+        return null;
+      }
+
+      logger.debug({ url: sshUrl.href }, `using ssh URL`);
+      return sshUrl.href;
+    }
+    case 'endpoint': {
+      const generatedUrl = generateUrlFromEndpoint(
+        defaultEndpoint,
+        opts,
+        repository
+      );
+      logger.debug({ url: generatedUrl }, `using endpoint URL`);
+      return generatedUrl;
+    }
+    case undefined: {
+      //if no connection type is specified try them all as a fallback
+      const httpUrl = getRepoGitUrl(
+        repository,
+        defaultEndpoint,
+        'default',
+        info,
+        opts
+      );
+      if (httpUrl !== null) {
+        return httpUrl;
+      }
+
+      const sshUrl = getRepoGitUrl(
+        repository,
+        defaultEndpoint,
+        'ssh',
+        info,
+        opts
+      );
+      if (sshUrl !== null) {
+        return sshUrl;
+      }
```

Comment:
```suggestion
): string {
  if (gitUrl === 'ssh') {
    const sshUrl = info.links.clone?.find(({ name }) => name === 'ssh');
    if (sshUrl === undefined) {
      throw new Error(CONFIG_GIT_URL_UNAVAILABLE);
    }

    logger.debug({ url: sshUrl.href }, `using ssh URL`);
    return sshUrl.href;
  }

  let cloneUrl = info.links.clone?.find(({ name }) => name === 'http');
  if (cloneUrl) {
    // Inject auth into the API provided URL
    return injectAuth(cloneUrl.href, opts);
  }

  // Http access might be disabled, try to find ssh url in this case
  cloneUrl = info.links.clone?.find(({ name }) => name === 'ssh');

  if (gitUrl === 'endpoint' || !cloneUrl) {
    return generateUrlFromEndpoint(defaultEndpoint, opts, repository);
  }

  // SSH urls can be used directly
  return cloneUrl.href;
}
```

Reply:
I am slightly concerned a user could set `gitUrl` to `endpoint` and it would be ignored if `http` or `ssh` are available. Is this the desired functionality?